### PR TITLE
BUGFIX: SD-943 Measure is removed when switch from geo chart to some charts

### PR DESCRIPTION
**Pull request by ptquang86**
---

[Check PR owner responsibilities](https://confluence.intgdc.com/display/Development/Code-reviews#Code-reviews-Ownerresponsibilities)

Supported PR commands:

Command | Description
--- | ---
`extended test - examples` | Live examples tests
`extended test - storybook` | Storybook screenshot tests

See more [options](https://confluence.intgdc.com/display/kbhr/How+to+work+with+git+and+Github#HowtoworkwithgitandGithub-Extendedchecks).

# PR Checklist

- [x] Verify code changes ([Checklist](https://confluence.intgdc.com/display/Development/Code-reviews+checklist), [Best practices](https://confluence.intgdc.com/display/Development/Code-reviews+best+practices))
- [x] [Verify pull-request formalities](https://confluence.intgdc.com/display/Development/Code-reviews)
- [x] Change was tested by using [gdc-dev-release](https://confluence.intgdc.com/display/~tomas.vojtasek/Private+NPM) in [gdc-analytical-designer](https://github.com/gooddata/gdc-analytical-designer) and [gdc-dashboards](https://github.com/gooddata/gdc-dashboards) (if applicable)
- [x] Migration guide (for major changes) is mentioned in [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [x] Successful `extended test - examples`
- [x] Successful `extended test - storybook`
- [x] Checked yarn.lock consistency (no dep. duplicities especially Goodstrap)
- [x] Squash commits

# Related PRs
- gdc-analytical-designer: https://github.com/gooddata/gdc-analytical-designer/pull/3097
- gdc-dashboards: https://github.com/gooddata/gdc-dashboards/pull/2812

# Related Jira tasks
- SD-943: https://jira.intgdc.com/browse/SD-943

### CODE REVIEW

**Comment by ptquang86** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7311/ : SUCCESS in 3m 28s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3320/ : SUCCESS in 3s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3833/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3321/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/3197/ : SUCCESS in 11m 15s

**Comment by ptquang86** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1599.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3325/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/3199/ : SUCCESS in 17m 32s

**Comment by ptquang86** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7390/ : SUCCESS in 2m 52s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3410/ : SUCCESS in 3s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3863/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3411/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/3219/ : SUCCESS in 9m 52s

**Comment by ptquang86** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1599.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3412/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/3220/ : SUCCESS in 16m 30s

File: src/internal/utils/bucketHelper.ts [hunk 1]
```diff
@@ -850,8 +850,9 @@ export const transformMeasureBuckets = (
 ) => {
     let unusedMeasures: IBucketItem[] = [];
 
-    const newBuckets: IBucket[] = measureBucketItemsLimits.map(({ localIdentifier, itemsLimit }) => {
-        const preferredBucketItems = getPreferredBucketItems(buckets, [localIdentifier], [METRIC]);
+    const newBuckets: IBucket[] = measureBucketItemsLimits.map(({ localIdentifiers, itemsLimit }) => {
+        const localIdentifier = localIdentifiers[0] || BucketNames.MEASURES;
```
Comment (naibinh) [inline src/internal/utils/bucketHelper.ts:854]: How about `const [localIdentifier = BucketNames.MEASURES] = localIdentifiers`?

Comment (ptquang86) [reply src/internal/utils/bucketHelper.ts]: Updated. Thank you

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7395/ : SUCCESS in 4m 16s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3418/ : SUCCESS in 6s

**Comment by naibinh** [comment]
LGTM

File: src/internal/utils/bucketHelper.ts [hunk 2]
```diff
@@ -850,8 +850,9 @@ export const transformMeasureBuckets = (
 ) => {
     let unusedMeasures: IBucketItem[] = [];
 
-    const newBuckets: IBucket[] = measureBucketItemsLimits.map(({ localIdentifier, itemsLimit }) => {
-        const preferredBucketItems = getPreferredBucketItems(buckets, [localIdentifier], [METRIC]);
+    const newBuckets: IBucket[] = measureBucketItemsLimits.map(({ localIdentifiers, itemsLimit }) => {
```
Comment (lanhq) [inline src/internal/utils/bucketHelper.ts:853]: how about if we rewrite this function as below? Then all refactoring on other files can be reverted.
```javascripts
export const transformMeasureBuckets = (
    measureBucketItemsLimits: IMeasureBucketItemsLimit[],
    buckets: IBucket[],
) => {
    let unusedMeasures: IBucketItem[] = [];
    const newBuckets: IBucket[] = measureBucketItemsLimits.map(({ localIdentifier, itemsLimit }) => {
    const preferedBucketlocalIdentifiers = localIdentifier === BucketNames.MEASURES
        ? [BucketNames.MEASURES, BucketNames.SIZE]
        : localIdentifier === BucketNames. SECONDARY_MEASURES
            ? [BucketNames.SECONDARY_MEASURES, BucketNames.COLOR]
            : [localIdentifier];
    const preferredBucketItems = getPreferredBucketItems(buckets, preferedBucketlocalIdentifiers, [METRIC]);
    const measuresToBePlaced = preferredBucketItems.splice(0, itemsLimit);

    if (measuresToBePlaced.length === 0) {
        return {
            localIdentifier,
            items: unusedMeasures.splice(0, itemsLimit),
        };
    }
    unusedMeasures = [...unusedMeasures, ...preferredBucketItems];
    return {
        localIdentifier,
        items: measuresToBePlaced,
    };
});
```

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7404/ : SUCCESS in 2m 53s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3429/ : SUCCESS in 3s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7405/ : SUCCESS in 3m 23s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3430/ : SUCCESS in 4s

File: src/internal/utils/bucketHelper.ts (hunk 2 above)
Comment (ptquang86) [reply src/internal/utils/bucketHelper.ts]: updated

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3864/ : SUCCESS in 16m 46s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-fossa-scanning-zuul-el7/23/ : SUCCESS in 17s (non-voting)
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7407/ : SUCCESS in 2m 38s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/1798/ : SUCCESS in 12s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/2643/ : FAILURE in 14m 13s

**Comment by ptquang86** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1599.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3432/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/3224/ : SUCCESS in 5m 31s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3865/ : SUCCESS in 13m 43s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-fossa-scanning-zuul-el7/24/ : SUCCESS in 17s (non-voting)
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7408/ : SUCCESS in 4m 37s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/1799/ : SUCCESS in 15s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/2645/ : SUCCESS in 8m 25s

**Pull request merged**